### PR TITLE
Add CSV Source

### LINKED ISSUE #52: Add CSV source
Blocked by #49 (This PR adds a source trait which we should implement for a CSV source)

### Description
We want to add an additional data source to the sync tool, specifically CSV (or some other form of flatfile source). We have a need for this quite soon, as it would significantly simplify the process of bootstrapping the demo environment for Operations and Marketing.

### User story
As a Famedly customer, I want to be able to provide a csv file of user details to import into Famedly so I can easily provision my list of users in bulk.

### Requirements
* CSV format (needs to also define the format so we can communicate this to Marketing who will use it, as well as document it in the readme)

### Steps to implement
* Add a new implementation of the source trait that reads a CSV file from disk

### PR DESCRIPTION
**Pull request by jannden**
Closes #52 , https://github.com/famedly/product-management/issues/2038

### CODE REVIEW

**Comment by jannden** [comment]
The requirements were clarified and confirmed with the only stakeholder at the moment - the Marketing team. Task specified https://github.com/famedly/ldap-sync/issues/52
Namely:
- only importing new users,
- skipping import for users that already exist,
- reading CSV file from disk,
- CSV data format: `email,first_name,last_name,preferred_username,phone,enabled`

File: src/sources/csv.rs [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+	fn read_csv(&self) -> Result<Vec<User>> {
+		let file_path = &self.csv_config.file_path;
+		let file = fs::File::open(&self.csv_config.file_path)
+			.context(format!("Failed to open CSV file {}", file_path.to_string_lossy()))?;
+		let mut reader = Reader::from_reader(file);
+		let users: Result<Vec<CsvData>, _> = reader.deserialize().collect();
+		let users: Result<Vec<User>, _> = users
+			.context("Failed to deserialize CSV data")
+			.map(|users| users.into_iter().map(CsvData::to_user).collect());
+		users
```
Comment (sirewix) [inline src/sources/csv.rs:47]: Small suggestion to avoid unnecessary bindings:
```rust
		Ok(reader
			.deserialize()
			.collect::<Result<Vec<_>, _>>()
			.context("Failed to deserialize CSV data")?
			.into_iter()
			.map(CsvData::to_user)
			.collect())
```
Or to actually ignore deserialization errors:
```rust
		Ok(reader
			.deserialize()
			.map(|r| r.inspect_err(|x| tracing::error!("Failed to deserialize: {x}")))
			.filter_map(Result::ok)
			.map(CsvData::to_user)
			.collect())
```
Or if you still want to keep it splitted, we should unwrap IMO instead of `.map` on `Result`:
```rust
		let users: Vec<CsvData> = reader
			.deserialize()
			.collect::<Result<Vec<_>, _>>()
			.context("Failed to deserialize CSV data")?;
		Ok(users.into_iter().map(CsvData::to_user).collect())
```

File: src/sources/csv.rs [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+			phone: if csv_data.phone.is_empty() { None } else { Some(csv_data.phone.into()) },
+			enabled: csv_data.enabled,
+			preferred_username: csv_data.preferred_username.clone().into(),
+			ldap_id: csv_data.preferred_username.into(),
+		}
+	}
+}
+
+/// Helper module for unit and e2e tests
+pub mod test_helpers {
```
Comment (sirewix) [inline src/sources/csv.rs:91]: `#[cfg(test)]` maybe?

File: src/sources/csv.rs [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+
+#[async_trait]
+impl Source for CsvSource {
+	fn get_name(&self) -> &'static str {
+		"CSV"
+	}
+
+	async fn get_diff(&self) -> Result<SourceDiff> {
+		let new_users = self.read_csv()?;
+		return Ok(SourceDiff { new_users, changed_users: vec![], deleted_user_ids: vec![] });
```
Comment (sirewix) [inline src/sources/csv.rs:27]: Continuing on [the past discussion](https://github.com/famedly/ldap-sync/pull/49#discussion_r1754790510), I think we should at least add a comment that this only implements adding users.

Potentially this could work for changing users as well, as for deleting it would require different csv table format

File: src/sources/csv.rs (hunk 2 above)
Comment (sirewix) [reply src/sources/csv.rs:91]: We should add `#![cfg(test)]` to those (as in https://github.com/famedly/backend-system-tests/blob/ef787dc3d6a882e18f1333abf9c5ee73e7935151/tests/zitadel.rs#L1)

File: src/sources/csv.rs (hunk 3 above)
Comment (tlater-famedly) [reply src/sources/csv.rs:27]: Agreed. I think we should also double check with marketing whether this is actually the *desired* behavior. Could well be that they didn't consider the edge case of "some existing user on an instance matches with the ones we want to sync but has outdated data".

Of course, if we want to update users we need to first query Zitadel, which is basically #53, and in fact updating users will make #53 significantly easier to make generic. We can merge this PR either way, but we should add a TODO if we want to change the behavior.

File: src/sources/csv.rs [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+#[derive(Debug, Deserialize)]
+struct CsvData {
+	/// The user's email address
+	email: String,
+	/// The user's first name
+	first_name: String,
+	/// The user's last name
+	last_name: String,
+	/// The user's preferred username
+	preferred_username: String,
```
Comment (tlater-famedly) [inline src/sources/csv.rs:68]: We basically only record this to add it to metadata to remember it since it's available on ldap, it's never used in practice. Do we actually need to define it for the CSV source?

File: src/sources/csv.rs [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+	/// The user's phone number
+	phone: String,
+	/// Whether the user is enabled
+	enabled: bool,
```
Comment (tlater-famedly) [inline src/sources/csv.rs:72]: This should always be true, disabled users are simply deleted from Zitadel

File: src/user.rs [hunk 6]
```diff
@@ -39,7 +39,7 @@ impl User {
 
 /// Crate-internal representation of a Zitadel user
 #[derive(Clone, Debug)]
-pub struct ZidatelUser {
+pub struct ZitadelUser {
```
Comment (tlater-famedly) [inline src/user.rs:42]: Let's put this in a separate commit, this just muddies the blame.

File: tests/environment/files/test-users.csv [hunk 7]
```diff
@@ -0,0 +1,5 @@
+email,first_name,last_name,preferred_username,phone,enabled
+john.doe@example.com,John,Doe,jdoe,"+1111111111",true
+jane.smith@example.com,Jane,Smith,jasmith,"+2222222222",true
+alice.johnson@example.com,Alice,Johnson,ajohnson,,true
+bob.williams@example.com,Bob,Williams,bwilliams,"+4444444444",true
```
Comment (tlater-famedly) [inline tests/environment/files/test-users.csv:5]: Nit: Another EOF newline ;) The precommit hook checks for this, maybe we should add a check to CI as well.

File: src/sources/csv.rs (hunk 4 above)
Comment (tlater-famedly) [reply src/sources/csv.rs:68]: May need to ask product for specifics here.

File: src/user.rs (hunk 6 above)
Comment (emgrav) [reply src/user.rs:42]: This is also already corrected on main

Comment (tlater-famedly) [reply src/user.rs:42]: Yep, spotted that after, ta @emgrav

File: src/sources/csv.rs (hunk 2 above)
Comment (jannden) [reply src/sources/csv.rs:91]: rustc doesn't like that and says `#![no_std]` above the test_helpers module and says inner attributes are usually found at the beginning of source files

File: src/sources/csv.rs (hunk 3 above)
Comment (jannden) [reply src/sources/csv.rs:27]: I added an explanatory comment.

File: src/sources/csv.rs (hunk 2 above)
Comment (sirewix) [reply src/sources/csv.rs:91]: I mean we can put `#![cfg(test)]` to the top of the `test/*.rs` modules, then they should compile if `test_helpers` is feature gated by `#[cfg(test)]`

File: config.sample.yaml [hunk 8]
```diff
@@ -107,3 +107,11 @@ sources:
...
     # Grant type
     grant_type: client_credentials
+
+  # Configuration for the CSV source - reads a CSV file
+  # and creates **new** users in Famedly's Zitadel.
+  # Expected structure of the CSV file is as follows:
+  # email,first_name,last_name,preferred_username,phone,enabled
+  csv:
+    # Path to the CSV file to read from.
+    file_path:  ./tests/environment/files/test-users.csv
```
[the remark this answers is not in the document]
Comment (jannden) [reply config.sample.yaml:117]: Done.

File: Cargo.toml [hunk 9]
```diff
@@ -25,6 +25,8 @@ zitadel-rust-client = { git = "https://github.com/famedly/zitadel-rust-client",
 chrono = "0.4.19"
```
[the remark this answers is not in the document]
Comment (jannden) [reply Cargo.toml:25]: Done.

File: src/sources/csv.rs (hunk 4 above)
Comment (jannden) [reply src/sources/csv.rs:68]: Yes, looks like it's kept for linking LDAP user to Zitadel user and it's quite entangled in the old code, specifically the User struct and everywhere it's used. I will see to what would be the best way to separate it. Probably making it an Option so that LDAP still can use it, but other sources don't have to.

File: tests/environment/files/test-users.csv (hunk 7 above)
Comment (jannden) [reply tests/environment/files/test-users.csv:5]: Done.

File: src/sources/csv.rs (hunk 2 above)
Comment (jannden) [reply src/sources/csv.rs:91]: The thing is that those modules have to be re-exported from lib.rs, which gives: found an item that was configured out. Am I missing something?

File: src/sources/csv.rs (hunk 5 above)
Comment (jannden) [reply src/sources/csv.rs]: Done.

File: src/sources/csv.rs (hunk 4 above)
Comment (jannden) [reply src/sources/csv.rs]: Done. I made the preferred_username being the email address by default if not supplied by LDAP otherwise.

File: src/sources/csv.rs (hunk 2 above)
Comment (sirewix) [reply src/sources/csv.rs:91]: you're right, that's impossible

**Comment by emgrav** [comment]
UKF wants a similar feature, apparently. See the linked issue above. I think it would make sense to add fields for organization, project, and role.

Sorry about not catching this issue earlier

**Comment by nikzen** [comment]
> UKF wants a similar feature, apparently. See the linked issue above. I think it would make sense to add fields for organization, project, and role.
> 
> Sorry about not catching this issue earlier

@emgrav THis is not required right now for UKF - so we can keep it as it is

**Review by sirewix — APPROVED** [review]

File: src/sources/csv.rs (hunk 4 above)
Comment (tlater-famedly) [reply src/sources/csv.rs]: *Technically* I think it'd be best if we didn't set this field at all when the CSV source is used. I'm ok with either implementation though, given we don't actually know why we're setting this field at the moment.

**Review by tlater-famedly — APPROVED** [review]

File: Cargo.toml [hunk 10]
```diff
@@ -26,6 +26,8 @@ url = "2.5.2"
 uuid = { version = "1.10.0", features = ["v5"] }
 zitadel-rust-client = { git = "https://github.com/famedly/zitadel-rust-client", version = "0.1.0" }
 wiremock = "0.6.2"
+csv = "1.3.0"
+tempfile = "3.12.0"
```
[the remark this answers is not in the document]
Comment (jannden) [reply Cargo.toml:30]: Looks like `tempfile` already was in [the dependencies](https://github.com/famedly/ldap-sync/blob/main/Cargo.lock#L1505). Weird historically, but all looks good at this point.

[the remark this answers is not in the document]
Comment (tlater-famedly) [reply Cargo.toml:30]: Fair enough, I assume it's a dependency of something.
